[ES] ... org.elasticsearch.index.query.QueryShardException: ... input automaton is too large: 1001

### Setup and configuration

```
{
    "SMWElasticStore": {
        "mysql": "10.1.37-MariaDB-0+deb9u1",
        "es": "6.6.0"
    },
    "smw": "3.1.0-alpha",
    "mediawiki": "1.32.0",
    "php": "7.0.33-0+deb9u1"
}
```

### Issue

https://sandbox.semantic-mediawiki.org/wiki/Discussion_utilisateur:Kghbln#Elastic_replication.3F

### Stack trace

```
[2019-02-04T05:46:37,800][DEBUG][o.e.a.s.TransportSearchAction] [Ql8wRCh] All shards failed for phase: [query]
org.elasticsearch.index.query.QueryShardException: failed to create query: {
  "constant_score" : {
    "filter" : {
      "bool" : {
        "must" : [
          {
            "range" : {
              "P:10.wpgField" : {
                "from" : "O:35:\"SMW\\DataModel\\ContainerSemanticData\":10:{s:11:\"\u0000*\u0000mSubject\";O:14:\"SMW\\DIWikiPage\":9:{s:10:\"\u0000*\u0000m_dbkey\";s:17:\"SMWInternalObject\";s:14:\"\u0000*\u0000m_namespace\";i:-1;s:14:\"\u0000*\u0000m_interwiki\";s:0:\"\";s:18:\"\u0000*\u0000m_subobjectname\";s:3:\"int\";s:23:\"\u0000SMW\\DIWikiPage\u0000sortkey\";N;s:32:\"\u0000SMW\\DIWikiPage\u0000contextReference\";N;s:28:\"\u0000SMW\\DIWikiPage\u0000pageLanguage\";N;s:18:\"\u0000SMW\\DIWikiPage\u0000id\";i:0;s:20:\"\u0000SMWDataItem\u0000options\";N;}s:14:\"\u0000*\u0000mProperties\";a:2:{s:6:\"_LCODE\";O:14:\"SMW\\DIProperty\":5:{s:21:\"\u0000SMW\\DIProperty\u0000m_key\";s:6:\"_LCODE\";s:25:\"\u0000SMW\\DIProperty\u0000m_inverse\";b:0;s:33:\"\u0000SMW\\DIProperty\u0000propertyValueType\";s:7:\"__lcode\";s:25:\"\u0000SMW\\DIProperty\u0000interwiki\";s:0:\"\";s:20:\"\u0000SMWDataItem\u0000options\";N;}s:5:\"_TEXT\";O:14:\"SMW\\DIProperty\":5:{s:21:\"\u0000SMW\\DIProperty\u0000m_key\";s:5:\"_TEXT\";s:25:\"\u0000SMW\\DIProperty\u0000m_inverse\";b:0;s:33:\"\u0000SMW\\DIProperty\u0000propertyValueType\";s:4:\"_txt\";s:25:\"\u0000SMW\\DIProperty\u0000interwiki\";s:0:\"\";s:20:\"\u0000SMWDataItem\u0000options\";N;}}s:12:\"\u0000*\u0000mPropVals\";a:2:{s:5:\"_TEXT\";a:1:{s:120:\"Used in cases where a URL property needs to be exported as an \"owl:AnnotationProperty\". It is a variant of datatype URL.\";O:9:\"SMWDIBlob\":2:{s:11:\"\u0000*\u0000m_string\";s:120:\"Used in cases where a URL property needs to be exported as an \"owl:AnnotationProperty\". It is a variant of datatype URL.\";s:20:\"\u0000SMWDataItem\u0000options\";N;}}s:6:\"_LCODE\";a:1:{s:2:\"en\";O:9:\"SMWDIBlob\":2:{s:11:\"\u0000*\u0000m_string\";s:2:\"en\";s:20:\"\u0000SMWDataItem\u0000options\";N;}}}s:19:\"\u0000*\u0000mHasVisibleProps\";b:1;s:19:\"\u0000*\u0000mHasVisibleSpecs\";b:1;s:16:\"\u0000*\u0000mNoDuplicates\";b:1;s:55:\"\u0000SMW\\DataModel\\ContainerSemanticData\u0000skipAnonymousCheck\";b:1;s:18:\"\u0000*\u0000subSemanticData\";O:29:\"SMW\\DataModel\\SubSemanticData\":2:{s:38:\"\u0000SMW\\DataModel\\SubSemanticData\u0000subject\";r:2;s:46:\"\u0000SMW\\DataModel\\SubSemanticData\u0000subSemanticData\";a:0:{}}s:10:\"\u0000*\u0000options\";N;s:16:\"\u0000*\u0000extensionData\";a:1:{s:9:\"sort.data\";s:123:\"Used in cases where a URL property needs to be exported as an \"owl:AnnotationProperty\". It is a variant of datatype URL.;en\";}}",
                "to" : null,
                "include_lower" : false,
                "include_upper" : true,
                "boost" : 1.0
              }
            }
          }
        ],
        "adjust_pure_negative" : true,
        "boost" : 1.0
      }
    },
    "boost" : 1.0
  }
}
	at org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:324) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:307) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:806) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:656) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:631) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:388) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService.access$100(SearchService.java:126) ~[elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:360) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:356) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.search.SearchService$4.doRun(SearchService.java:1117) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:759) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) [elasticsearch-6.6.0.jar:6.6.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.6.0.jar:6.6.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_181]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
Caused by: java.lang.IllegalArgumentException: input automaton is too large: 1001
	at org.apache.lucene.util.automaton.Operations.isFinite(Operations.java:1037) ~[lucene-core-7.6.0.jar:7.6.0 719cde97f84640faa1e3525690d262946571245f - nknize - 2018-12-07 14:44:20]
    ...
```

### Steps to reproduce

Unknown!

I don't know why we have a range query for `P:10.wpgField` (`_PDESC` property) and who emitted this query but one action point could be to restrict the value length for a search request to avoid a potential "Caused by: java.lang.IllegalArgumentException: input automaton is too large: 1001" [0] to 500 chars as max length.

[0] https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/util/automaton/Operations.java#L1037

--- a/DefaultSettings.php
+++ b/DefaultSettings.php
@@ -2137,6 +2137,12 @@ return [
 			// when resolved in connection with a query
 			'debug.description.log' => true,
 
+			// #3698
+			// Restrict the length of an individual input value to avoid a potential
+			// "... "java.lang.IllegalArgumentException:input automaton is too
+			// large: 1001 ..."
+			'maximum.value.length' => 500,
+
 			// When `!...` is used make sure that the condition is only applied
 			// on entities where the property exists together with the negated
 			// value condition otherwise the ! condition becomes unrestricted

--- a/src/Elastic/QueryEngine/DescriptionInterpreters/SomeValueInterpreter.php
+++ b/src/Elastic/QueryEngine/DescriptionInterpreters/SomeValueInterpreter.php
@@ -145,6 +145,10 @@ class SomeValueInterpreter {
 			$value = $dataItem->getSortKey();
 		}
 
+		if ( mb_strlen( $value ) > $this->conditionBuilder->getOption( 'maximum.value.length' ) ) {
+			$value = mb_substr( $value, 0, $this->conditionBuilder->getOption( 'maximum.value.length' ) );
+		}
+
 		if ( $this->isRange( $comparator ) ) {
 			$match = $this->fieldMapper->range( "$pid.$field", $value, $comparator );
 		} elseif ( $isSubDataType && $dataItem->getDBKey() === '' && $comparator === SMW_CMP_NEQ ) {
@@ -271,6 +275,10 @@ class SomeValueInterpreter {
 		$type = $options['type'];
 
 		$value = $dataItem->getSerialization();
+
+		if ( mb_strlen( $value ) > $this->conditionBuilder->getOption( 'maximum.value.length' ) ) {
+			$value = mb_substr( $value, 0, $this->conditionBuilder->getOption( 'maximum.value.length' ) );
+		}
 
 		if ( $this->isRange( $comparator ) ) {
 			// Use a not_analyzed field
@@ -440,6 +448,10 @@ class SomeValueInterpreter {
 	 * @return array
 	 */
 	public function plain( $value, array &$options ) {
+
+		if ( mb_strlen( $value ) > $this->conditionBuilder->getOption( 'maximum.value.length' ) ) {
+			$value = mb_substr( $value, 0, $this->conditionBuilder->getOption( 'maximum.value.length' ) );
+		}
 
 		$comparator = $options['comparator'];
 		$pid = $options['pid'];

--- a/src/Elastic/QueryEngine/DescriptionInterpreters/ValueDescriptionInterpreter.php
+++ b/src/Elastic/QueryEngine/DescriptionInterpreters/ValueDescriptionInterpreter.php
@@ -108,6 +108,10 @@ class ValueDescriptionInterpreter {
 			$value = $dataItem->getSerialization();
 		}
 
+		if ( mb_strlen( $value ) > $this->conditionBuilder->getOption( 'maximum.value.length' ) ) {
+			$value = mb_substr( $value, 0, $this->conditionBuilder->getOption( 'maximum.value.length' ) );
+		}
+
 		if ( $dataItem instanceof DIWikiPage && $this->isRange( $comparator ) ) {
 			$params = $this->fieldMapper->range( "$field.keyword", $value, $comparator );
 		} elseif ( $dataItem instanceof DIBlob && $comparator === SMW_CMP_EQ ) {

--- a/tests/phpunit/Unit/Elastic/QueryEngine/DescriptionInterpreters/SomeValueInterpreterTest.php
+++ b/tests/phpunit/Unit/Elastic/QueryEngine/DescriptionInterpreters/SomeValueInterpreterTest.php
@@ -35,6 +35,12 @@ class SomeValueInterpreterTest extends \PHPUnit_Framework_TestCase {
 			->disableOriginalConstructor()
 			->setMethods( [ 'getID' ] )
 			->getMock();
+
+		$this->conditionBuilder->setOptions( new Options(
+			[
+				'maximum.value.length' => 500
+			]
+		) );
 	}
 
 	public function testCanConstruct() {
@@ -168,7 +174,8 @@ class SomeValueInterpreterTest extends \PHPUnit_Framework_TestCase {
 
 		$this->conditionBuilder->setOptions( new Options(
 			[
-				'cjk.best.effort.proximity.match' => true
+				'cjk.best.effort.proximity.match' => true,
+				'maximum.value.length' => 500
 			]
 		) );
 

--- a/tests/phpunit/Unit/Elastic/QueryEngine/DescriptionInterpreters/ValueDescriptionInterpreterTest.php
+++ b/tests/phpunit/Unit/Elastic/QueryEngine/DescriptionInterpreters/ValueDescriptionInterpreterTest.php
@@ -51,7 +51,8 @@ class ValueDescriptionInterpreterTest extends \PHPUnit_Framework_TestCase {
 
 		$this->conditionBuilder->setOptions( new Options(
 			[
-				'cjk.best.effort.proximity.match' => true
+				'cjk.best.effort.proximity.match' => true,
+				'maximum.value.length' => 500
 			]
 		) );
 
@@ -72,6 +73,43 @@ class ValueDescriptionInterpreterTest extends \PHPUnit_Framework_TestCase {
 
 		$this->assertEquals(
 			$expected,
+			(string)$condition
+		);
+	}
+
+	public function testRestrictedLength() {
+
+		$options = [
+			'property' => $this->dataItemFactory->newDIProperty( 'Bar' ),
+			'pid'   => 'P:42',
+			'field' => 'wpgID',
+			'type'  => 'must'
+		];
+
+		$this->conditionBuilder->setOptions( new Options(
+			[
+				'maximum.value.length' => 1
+			]
+		) );
+
+		$instance = new ValueDescriptionInterpreter(
+			$this->conditionBuilder
+		);
+
+		$description = $this->descriptionFactory->newValueDescription(
+			$this->dataItemFactory->newDIWikiPage( 'test' ),
+			null,
+			SMW_CMP_EQ
+		);
+
+		$condition = $instance->interpretDescription(
+			$description,
+			$options
+		);
+
+		// 4 vs. 42
+		$this->assertEquals(
+			'{"bool":{"must":{"terms":{"_id":["4"]}}}}',
 			(string)$condition
 		);
 	}

